galsim/detectors.py

```diff
@@ -275,6 +275,59 @@ def applyIPC(self, IPC_kernel, edge_treatment='extend', fill_value=None, kernel_
     else:
         self.array[:,:] = out_array
 
+def addPersistence(self,imgs,coeffs):
+    """
+    Adds the effects of persistence to the Image instance.
+
+    Persistence refers to the retention of a small fraction of the signal after resetting the
+    imager pixel elements. The persistence signal of a previous exposure is left in the pixel even
+    after several detector resets. This effect is most likely due to charge traps in the material.
+    The laboratory tests show that if exposures and readouts are taken in a fixed cadence, the
+    persistence signal can be given as a linear combination of prior pixel values that can be
+    added to the current image.
+
+    This routine takes in one or multiple Image instances and adds them to Image weighted by the
+    values passed on to 'coeffs'. This routine does NOT keep track of realistic dither patterns.
+    During the image simulation process, the user has to queue a list of previous Image
+    instances (img_list) outside the routine by inserting the latest image in the beginning of the
+    list and deleting the oldest image. The values in 'coeffs' tell how much of each Image is to
+    be added. This usually remains constant in the image generation process.
+
+    Calling
+    -------
+
+        >>> img.addPersistence(imgs=img_list, coeffs=coeffs_list)
+
+    @ param imgs:       an Image instance or a list of previous Image instances that still persist
+    @ param coeffs:     specifies the retention factor for one or all of the Image instances listed
+                        in 'imgs'. Different retention factors for different instances can be
+                        specified as a list or tuples in the same order as that of the images.
+
+    @ returns None
+    """
+
+    if isinstance(imgs,list) and hasattr(coeffs,'__iter__'):
+        if not len(imgs)==len(coeffs):
+            raise TypeError("The length of 'imgs' and 'coeffs' must be the same, if passed as a \
+                list")
+            # If this error is not raised, then the images are added as long as one of the list is
+            # exhausted.
+        else:
+            for img,coeff in zip(imgs,coeffs):
+                self += coeff*img
+
+    elif isinstance(imgs,list) and (isinstance(coeffs,float) or isinstance(coeffs,int)):
+        for img in imgs:
+            self += coeffs*img
```

Comment:
I thought about this before pushing the code. A more sensible way to treat
a float input for `coeffs` would be use the powers of the input value for
the past images - i.e. 0.01 for the previous exposure, 0.0001 for the
second previous exposure so on. But no where else in the routine do we
specify the order of the images. So I am up for removing this case totally.

On Wed, Feb 10, 2016 at 2:00 PM, Mike Jarvis <notifications@github.com>
wrote:

> In galsim/detectors.py
> <https://github.com/GalSim-developers/GalSim/pull/710#discussion_r52506279>
> :
>
> > +    @ returns None
> > +    """
> > +
> > +    if isinstance(imgs,list) and hasattr(coeffs,'__iter__'):
> > +        if not len(imgs)==len(coeffs):
> > +            raise TypeError("The length of 'imgs' and 'coeffs' must be the same, if passed as a \
> > +                list")
> > +            # If this error is not raised, then the images are added as long as one of the list is
> > +            # exhausted.
> > +        else:
> > +            for img,coeff in zip(imgs,coeffs):
> > +                self += coeff*img
> > +
> > +    elif isinstance(imgs,list) and (isinstance(coeffs,float) or isinstance(coeffs,int)):
> > +        for img in imgs:
> > +            self += coeffs*img
>
> Is this case relevant? If we have multiple images, I think we ought to
> require a list of coefficients. If the user supplies a float for coeffs, it
> seems to me that it would more likely be an error than that they intended
> for this branch of the code.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GalSim-developers/GalSim/pull/710/files#r52506279>.
>



-- 
Regards

Arun Kannawadi Jayaraman
PhD candidate, Dept. of Physics,
Carnegie Mellon University,
Pittsburgh, PA 15213